pocs/linux/kernelctf/CVE-2024-26808_cos/docs/exploit.md

```diff
@@ -0,0 +1,404 @@
+                goto nla_put_failure;
+            n++;
+        }
+        nla_nest_end(skb, nest_devs);
+
+        if (n == 1 &&
+            nla_put_string(skb, NFTA_HOOK_DEV, first->ops.dev->name))
+            goto nla_put_failure;
+```
+
+### Arbitrary free
+
+When we delete the chain, it will call `__nf_unregister_net_hook` to find out the `struct nf_hook_entries` by freed net_device and later call `nf_hook_entries_free` to free our controllable value. But there is a limit, we need to make our crafted net_device's net to be the same net in line 2. Otherwise it will return NULL and early return without go into `nf_hook_entries_free`
+
+```C
+static struct nf_hook_entries __rcu **
+nf_hook_entry_head(struct net *net, int pf, unsigned int hooknum,
+           struct net_device *dev)
+{
+
+...
+    case NFPROTO_INET:
+        if (WARN_ON_ONCE(hooknum != NF_INET_INGRESS))
+            return NULL;
+        if (!dev || dev_net(dev) != net) { //[2]
+            WARN_ON_ONCE(1);
+            return NULL;
+        }
+        return &dev->nf_hooks_ingress;
+```
+
+### Overwrite core_pattern
+One of methods to get privileged execution is overwrite `core_pattern` so our exploit will executed as root if any crash in user program is triggered. Using arbitrary free before, the exploit is free some `pipe_buffer` object and overwrite `pipe_buffer` with our own value.
+
+In `pipe_buffer` object there's `page` member that represents any allocated physical page in memory. To know where `core_pattern`'s page object is, first we spill struct page of some kernel text to the `pipe_buffer`, the trick we do is by `vmsplice` vDSO address to the pipe. Then, by reading `page` from victim `pipe_buffer` we can calculate the `struct page` of `core_pattern` is located. We also overwrite `flags` with `PIPE_BUF_FLAG_CAN_MERGE`, so by writing to the pipe we can overwrite the content of the memory page, which is `core_pattern`.
+
+### Root shell
+After overwrite `core_pattern`, we can crash the program, and anything in `core_pattern` will be executed as root which is our root shell.
+
+### Detailed summary of step by steps exploitation
+With these two limited primitive, our exploit is in the following steps:
+
+1. Cross cache to overwrite `net_device` object in kmalloc-cg-4k to `packet_fanout` object in kmalloc-4k. 
+2. `packet_fanout` object has `net` address in the first 8 byte, we leak this using UAF leak primitive.
+3. Using another UAF object, this time we overwrite `net_device` with `msg_msg` object, no need cross cache we use same kmalloc-cg-4k.
+4. We fill `msg_msg.next` (first 8 byte offset) with kmalloc-cg-192 by send `msg_msg` with size 0x90 for every `msg_msg` that already placed in kmalloc-cg-4k.
+5. Leak kmalloc-cg-192 `msg_msg` addr using UAF leak.
+6. Use arbitrary free to free a chunk in kmalloc-cg-192 we got before.
+7. Reclaim it as `pipe_buffer` A and use arbitrary free again at same address.
+8. Reclaim it as another `pipe_buffer` B.
```

Comment:
Remove this step if not needed (from the documentation and from the exploit).

Reply:
fixed